syntax = "proto3";

import "google/protobuf/timestamp.proto";

service CommWithReplica{
    rpc SendDetailsOfPR(Address) returns (StatusRepReq) {}
    rpc Write(WriteRequest) returns (WriteResponse) {}
    rpc Read(ReadRequest) returns (ReadResponse) {}
    rpc Delete(DeleteRequest) returns (StatusRepReq) {}
    rpc ConnectToPRforWrite(WriteRequest) returns (StatusRepReqWithVersion) {}
    rpc ConnectToPRforDelete(DeleteRequest) returns (StatusRepReq) {}
    rpc ConnectToReplicaforWrite(Request) returns (StatusRepReq) {}
    rpc ConnectToReplicaforDelete(DeleteRequestToReplica) returns (StatusRepReq) {}
}

message Address{
    string name = 1;
    string ip = 2;
    int32 port = 3;
}

message StatusRepReq{
    string status = 1;
}

message StatusRepReqWithVersion{
    string status = 1;
    google.protobuf.Timestamp version = 2;
}

message WriteRequest{
    string uuid = 1;
    string name = 2;
    string content = 3;
}

message Request{
    string uuid = 1;
    string name = 2;
    string content = 3;
    google.protobuf.Timestamp version = 4;
}

message ReadRequest{
    string uuid = 1;
}

message DeleteRequest{
    string uuid = 1;
}

message DeleteRequestToReplica{
    string uuid = 1;
    google.protobuf.Timestamp version = 2;
}

message WriteResponse{
    string status = 1;
    string uuid = 2;
    google.protobuf.Timestamp version = 3;
}

message ReadResponse{
    string status = 1;
    string name = 2;
    string content = 3;
    google.protobuf.Timestamp version = 4;
}
